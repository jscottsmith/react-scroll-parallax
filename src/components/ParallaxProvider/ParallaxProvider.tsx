import React, { Component } from 'react';

import { ParallaxContext } from '../../context/ParallaxContext';
import { ParallaxController, ScrollAxis } from 'parallax-controller';
import { ParallaxProviderProps } from './types';
import { createController } from './helpers';

export class ParallaxProvider extends Component<ParallaxProviderProps, {}> {
  static defaultProps = {
    scrollAxis: ScrollAxis.vertical,
  };

  controller: ParallaxController | null;

  constructor(props: ParallaxProviderProps) {
    super(props);
    this.controller = createController({
      scrollAxis: props.scrollAxis,
      scrollContainer: props.scrollContainer,
      disabled: props.isDisabled,
    });
  }

  componentDidUpdate(prevProps: ParallaxProviderProps) {
    if (
      prevProps.scrollContainer !== this.props.scrollContainer &&
      this.props.scrollContainer
    ) {
      this.controller?.updateScrollContainer(this.props.scrollContainer);
    }

    if (prevProps.isDisabled !== this.props.isDisabled) {
      if (this.props.isDisabled) {
        this.controller?.disableParallaxController();
      }
      if (!this.props.isDisabled) {
        this.controller?.enableParallaxController();
      }
    }
  }

  componentWillUnmount() {
    // @ts-ignore
    this.controller = this.controller.destroy();
  }

  render() {
    const { children } = this.props;
    return (
      // @ts-ignore
      <ParallaxContext.Provider value={this.controller}>
        {children}
      </ParallaxContext.Provider>
    );
  }
}
