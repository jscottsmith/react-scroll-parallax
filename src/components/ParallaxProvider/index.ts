export { ParallaxProvider } from './ParallaxProvider';
export * from './types';
