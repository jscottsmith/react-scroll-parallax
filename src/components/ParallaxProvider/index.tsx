import React, { Component } from 'react';

import { ParallaxContext } from '../../context/ParallaxContext';
import { ParallaxController, ScrollAxis } from 'parallax-controller';
import { ParallaxProviderProps } from './types';
import { createController } from './helpers';

export class ParallaxProvider extends Component<ParallaxProviderProps, {}> {
  static defaultProps = {
    scrollAxis: ScrollAxis.vertical,
  };

  controller: ParallaxController | null;

  constructor(props: ParallaxProviderProps) {
    super(props);
    this.controller = createController({
      scrollAxis: props.scrollAxis,
      scrollContainer: props.scrollContainer,
    });
  }

  componentDidUpdate(prevProps: ParallaxProviderProps) {
    if (prevProps.scrollContainer !== this.props.scrollContainer) {
      // @ts-ignore
      this.controller.updateScrollContainer(this.props.scrollContainer);
    }
  }

  componentWillUnmount() {
    // @ts-ignore
    this.controller = this.controller.destroy();
  }

  render() {
    const { children } = this.props;
    return (
      // @ts-ignore
      <ParallaxContext.Provider value={this.controller}>
        {children}
      </ParallaxContext.Provider>
    );
  }
}
