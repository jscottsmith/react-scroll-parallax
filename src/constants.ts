export const VERTICAL = 'vertical';
export const HORIZONTAL = 'horizontal';
