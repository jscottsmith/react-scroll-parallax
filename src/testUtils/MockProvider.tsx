import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ParallaxContext } from '../../src/context/ParallaxContext';
import { ParallaxController } from 'parallax-controller';

export class MockProvider extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    controllerMock: PropTypes.object.isRequired,
  };
  controller: ParallaxController;

  constructor(props: { controllerMock: ParallaxController }) {
    super(props);
    this.controller = props.controllerMock;
  }

  componentWillUnmount() {
    if (!this.controller) return;
    this.controller.destroy();
    // @ts-ignore
    this.controller = undefined;
  }

  render() {
    const { children } = this.props;

    return (
      <ParallaxContext.Provider value={this.controller}>
        {children}
      </ParallaxContext.Provider>
    );
  }
}
