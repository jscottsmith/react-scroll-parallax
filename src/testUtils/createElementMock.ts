export function createElementMock(properties = {}, methods = {}) {
  const element = document.createElement('div');
  Object.keys(properties).forEach((key) => {
    Object.defineProperty(element, key, {
      // @ts-ignore
      value: properties[key],
      writable: false,
    });
  });

  Object.keys(methods).forEach((key) => {
    // @ts-ignore
    element[key] = methods[key];
  });

  return element;
}
