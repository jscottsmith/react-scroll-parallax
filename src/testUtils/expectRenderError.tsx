/* eslint-disable */
const React = require('react');
const ReactDOM = require('react-dom');

export default function expectRenderError(
  element: any,
  expectedError: string,
  expectedErrorLength: number = 1
) {
  // Noop error boundary for testing.
  class TestBoundary extends React.Component {
    constructor(props: any) {
      super(props);
      this.state = { didError: false };
    }
    componentDidCatch() {
      this.setState({ didError: true });
    }
    render() {
      return this.state.didError ? null : this.props.children;
    }
  }

  // Record all errors.
  let topLevelErrors: ErrorEvent[] = [];
  function handleTopLevelError(event: ErrorEvent) {
    topLevelErrors.push(event.error);
    // Prevent logging
    event.preventDefault();
  }

  const div = document.createElement('div');
  window.addEventListener('error', handleTopLevelError);
  try {
    // @ts-ignore
    ReactDOM.render(<TestBoundary>{element}</TestBoundary>, div);
  } finally {
    window.removeEventListener('error', handleTopLevelError);
  }

  expect(topLevelErrors.length).toBe(expectedErrorLength);
  expect(topLevelErrors[0].message).toContain(expectedError);
}
