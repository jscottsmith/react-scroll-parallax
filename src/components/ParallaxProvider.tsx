import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ParallaxContext from '../helpers/ParallaxContext';
import ParallaxController from '../classes/ParallaxController';
import { VERTICAL, HORIZONTAL } from '../constants';
import validHTMLElement from '../utils/validHTMLElement';

const createController = (options) => {
  // Don't initialize on the server
  const isServer = typeof window === 'undefined';

  if (!isServer) {
    // Must not be the server so kick it off...
    return ParallaxController.init(options);
  }
  return null;
};

export interface ParallaxProviderProps {
  /**
   * Optionally pass the scroll axis for setting horizontal/vertical scrolling. One of vertical or
   * horizontal
   */
  scrollAxis?: 'vertical' | 'horizontal';
  /**
   * Optionally set the container that has overflow and will contain parallax elements. Defaults
   * to the HTML body
   */
  scrollContainer?: any;
}

export default class ParallaxProvider extends Component<
  ParallaxProviderProps,
  {}
> {
  static defaultProps = {
    scrollAxis: VERTICAL,
  };

  static propTypes = {
    children: PropTypes.node.isRequired,
    scrollAxis: PropTypes.oneOf([VERTICAL, HORIZONTAL]),
    scrollContainer: validHTMLElement,
  };

  constructor(props) {
    super(props);
    // @ts-ignore
    this.controller = createController({
      scrollAxis: props.scrollAxis,
      scrollContainer: props.scrollContainer,
    });
  }

  componentDidUpdate(prevProps) {
    if (prevProps.scrollContainer !== this.props.scrollContainer) {
      // @ts-ignore
      this.controller.updateScrollContainer(this.props.scrollContainer);
    }
  }

  componentWillUnmount() {
    // @ts-ignore
    this.controller = this.controller.destroy();
  }

  render() {
    const { children } = this.props;
    return (
      // @ts-ignore
      <ParallaxContext.Provider value={this.controller}>
        {children}
      </ParallaxContext.Provider>
    );
  }
}
