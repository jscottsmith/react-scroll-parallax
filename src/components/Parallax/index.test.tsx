import React, { PropsWithChildren } from 'react';
import {
  EffectNumber,
  ParallaxController,
  ScrollAxis,
} from 'parallax-controller';
import { render } from '@testing-library/react';
import { Parallax } from '.';
import { ParallaxProvider } from '../ParallaxProvider';
import { MockProvider } from '../../testUtils/MockProvider';
import expectRenderError from '../../testUtils/expectRenderError';
import { ALL_PARALLAX_PROPS } from '../../testUtils/tests.constants';

describe('given the <Parallax> component', () => {
  const preventError = (e: ErrorEvent) => e.preventDefault();

  beforeEach(() => {
    window.addEventListener('error', preventError);
  });

  afterEach(() => {
    window.removeEventListener('error', preventError);
  });

  describe('when normal configurations are given', () => {
    it('then it renders correctly', () => {
      const { asFragment } = render(
        <ParallaxProvider>
          <Parallax
            className="class-foo"
            disabled={false}
            translateX={[-100, 100]}
            translateY={['-100%', '100%']}
            style={{
              border: 'solid red 2px',
            }}
          >
            <div className="foo" />
          </Parallax>
        </ParallaxProvider>
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when given children', () => {
    it('then it renders them', () => {
      const { getByTestId } = render(
        <ParallaxProvider>
          <Parallax>
            <div data-testid="children" />
          </Parallax>
        </ParallaxProvider>
      );
      expect(getByTestId('children')).toBeInTheDocument();
    });
  });

  describe('when given html attributes', () => {
    it('then it adds them to the returned div', () => {
      const { container, getByTestId } = render(
        <ParallaxProvider>
          <Parallax
            style={{ background: 'red' }}
            className="my-class"
            id="test-id"
            data-testid="data-test-id"
            data-foo="bar"
            aria-label="Cool"
          />
        </ParallaxProvider>
      );
      expect(getByTestId('data-test-id')).toBeInTheDocument();
      expect(container.querySelector('.my-class')).toBeInTheDocument();
      expect(container.querySelector('#test-id')).toBeInTheDocument();
      expect(getByTestId('data-test-id')).toHaveAttribute('aria-label', 'Cool');
      expect(getByTestId('data-test-id')).toHaveAttribute('data-foo', 'bar');
      expect(getByTestId('data-test-id').style.background).toBe('red');
    });
  });

  describe('when not wrapped in the ParallaxProvider', () => {
    it('to throw because the ParallaxController is not available', () => {
      expectRenderError(
        <Parallax>
          <div />
        </Parallax>,
        'Could not find `react-scroll-parallax` context value. Please ensure the component is wrapped in a <ParallaxProvider>',
        1
      );
    });
  });

  describe.each(ALL_PARALLAX_PROPS)('when the prop %s is given', (props) => {
    it('then it renders without issue and calls create element with props', () => {
      const controller = ParallaxController.init({
        scrollAxis: ScrollAxis.vertical,
      });
      controller.createElement = jest.fn(controller.createElement);
      function Wrapper(props: PropsWithChildren<{}>) {
        return (
          <MockProvider controllerMock={controller}>
            {props.children}
          </MockProvider>
        );
      }
      const { asFragment } = render(<Parallax {...props} />, {
        wrapper: Wrapper,
      });
      expect(asFragment()).toMatchSnapshot();
      expect(controller.createElement).toBeCalledWith({
        el: expect.any(HTMLElement),
        props,
      });
    });
  });

  describe('when wrapped in the ParallaxProvider', () => {
    it('then is creates an element in the controller on mount', () => {
      const controller = ParallaxController.init({
        scrollAxis: ScrollAxis.vertical,
      });
      controller.createElement = jest.fn(controller.createElement);
      render(
        <MockProvider controllerMock={controller}>
          <Parallax translateY={[-100, 100]} />
        </MockProvider>
      );
      expect(controller.createElement).toBeCalledWith({
        el: expect.any(HTMLElement),
        props: { translateY: [-100, 100] },
      });
    });

    it('then it removes an element in the controller when unmounting', () => {
      const controller = ParallaxController.init({
        scrollAxis: ScrollAxis.vertical,
      });
      controller.removeElementById = jest.fn();

      const { unmount } = render(
        <MockProvider controllerMock={controller}>
          <Parallax>
            <div />
          </Parallax>
        </MockProvider>
      );
      const element = controller.getElements()[0];
      unmount();
      expect(controller.removeElementById).toBeCalledWith(element.id);
    });

    it('then it updates an element in the controller when receiving relevant new props', () => {
      const controller = ParallaxController.init({
        scrollAxis: ScrollAxis.vertical,
      });
      controller.updateElementPropsById = jest.fn();

      function Wrapper(props: PropsWithChildren<{}>) {
        return (
          <MockProvider controllerMock={controller}>
            {props.children}
          </MockProvider>
        );
      }

      const { rerender } = render(<Parallax disabled={true} />, {
        wrapper: Wrapper,
      });

      rerender(
        <Parallax
          disabled={false}
          translateX={[100, -100]}
          translateY={[-100, 100]}
        />
      );

      const element = controller.getElements()[0];

      expect(controller.updateElementPropsById).toBeCalledWith(element.id, {
        disabled: false,
        translateX: [100, -100],
        translateY: [-100, 100],
      });

      const newProps = {
        disabled: false,
        x: [-40, -60] as EffectNumber,
        y: [10, 80] as EffectNumber,
      };

      rerender(
        <Parallax
          disabled={newProps.disabled}
          translateX={newProps.x}
          translateY={newProps.y}
        />
      );

      expect(controller.updateElementPropsById).toBeCalledWith(element.id, {
        disabled: false,
        translateX: [-40, -60],
        translateY: [10, 80],
      });

      // only update with valid props
      rerender(
        <Parallax
          disabled={newProps.disabled}
          translateX={newProps.x}
          translateY={newProps.y}
          // @ts-expect-error
          nope="bad"
        />
      );

      expect(controller.updateElementPropsById).toHaveBeenCalledTimes(2);
    });

    it('then it resets styles on an element if the disabled prop is true', () => {
      const controller = ParallaxController.init({
        scrollAxis: ScrollAxis.vertical,
      });
      controller.resetElementStyles = jest.fn();

      function Wrapper(props: PropsWithChildren<{}>) {
        return (
          <MockProvider controllerMock={controller}>
            {props.children}
          </MockProvider>
        );
      }

      const offX: EffectNumber = [100, -100];
      const offY: EffectNumber = [100, -100];

      const { rerender } = render(
        <Parallax disabled={false} translateX={offX} translateY={offY} />,
        {
          wrapper: Wrapper,
        }
      );

      rerender(
        <Parallax disabled={true} translateX={offX} translateY={offY} />
      );

      expect(controller.resetElementStyles).toBeCalled();
    });
  });
});
