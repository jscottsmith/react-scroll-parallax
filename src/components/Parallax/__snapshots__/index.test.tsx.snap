// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given the <Parallax> component when normal configurations are given then it renders correctly 1`] = `
<DocumentFragment>
  <div
    class="class-foo"
    style="border: 2px solid red; will-change: transform; transform: translateX(100%)translateY(100%);"
  >
    <div
      class="foo"
    />
  </div>
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { disabled: true } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { easing: 'easeInQuad' } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { easing: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { onChange: [Function: onChange] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { onEnter: [Function: onEnter] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { onExit: [Function: onExit] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { onProgressChange: [Function: onProgressChange] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { opacity: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform,opacity; opacity: 1;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { rootMargin: [Object] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { rotate: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform; transform: rotate(1turn);"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { rotateX: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform; transform: rotateX(1turn);"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { rotateY: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform; transform: rotateY(1turn);"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { rotateZ: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform; transform: rotateZ(1turn);"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { scale: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform; transform: scale(1);"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { scaleX: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform; transform: scaleX(1);"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { scaleY: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform; transform: scaleY(1);"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { scaleZ: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform; transform: scaleZ(1);"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { shouldAlwaysCompleteAnimation: true } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { shouldDisableScalingTranslations: true } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { speed: 10 } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform; transform: translateY(-100px);"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { startScroll: 0, endScroll: 1000 } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { targetElement: HTMLDivElement {} } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform;"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { translateX: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform; transform: translateX(100%);"
  />
</DocumentFragment>
`;

exports[`given the <Parallax> component when the prop { translateY: [Array] } is given then it renders without issue and calls create element with props 1`] = `
<DocumentFragment>
  <div
    style="will-change: transform; transform: translateY(100%);"
  />
</DocumentFragment>
`;
